stablehlo/transforms/StablehloLegalizeQDQToQuantizedOp.cpp

```diff
@@ -0,0 +1,137 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"  // Include for TypeConverter
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOLEGALIZEQDQTOQUANTIZEDOPPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+bool isAnyQuantizedTypes(TypeRange types) {
+  return llvm::any_of(types, [](Type type) {
+    return isa<quant::QuantizedType>(getElementTypeOrSelf(type));
+  });
+}
+
+struct QuantizedStablehloQDQToQuantizedOpConversion
+    : public OpRewritePattern<stablehlo::UniformQuantizeOp> {
+  using OpRewritePattern<stablehlo::UniformQuantizeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(stablehlo::UniformQuantizeOp quantOp,
+                                PatternRewriter& rewriter) const override {
+    // Matching sequence of ops:
+    //    UniformDequantizeOp -> non-quantized Op -> UniformQuantizeOp.
+    // Start matching from a UniformQuantizeOp (`quantOp`).
+    // Get the Op (`computeOp`) which defines the inputs of the `quantOp`.
+    // Verify all inputs of the `computeOp` are produced by UniformDequantizeOp.
+    // Note: The pass does not delete any prexisting op.
+    auto* computeOp = quantOp->getOperand(0).getDefiningOp();
+    if (!computeOp)
+      return rewriter.notifyMatchFailure(
+          quantOp, "requires operand to be defined by an op");
+
+    if (computeOp->getNumRegions() != 0)
+      return rewriter.notifyMatchFailure(computeOp,
+                                         "ops with regions are not supported");
+
+    if (computeOp->getNumResults() > 1)
+      return rewriter.notifyMatchFailure(
+          computeOp, "ops with variadic results are not supported");
+
+    if (isAnyQuantizedTypes(computeOp->getOperandTypes()))
+      return rewriter.notifyMatchFailure(computeOp,
+                                         "requires non quantized operands");
+
+    // Collect quantized operands and result types to rewrite
+    // All operands and results must be quantized
+    llvm::SmallVector<Value> quantizedComputeOpOperands;
+    for (const Value& operand : computeOp->getOperands()) {
+      auto* definingOp = operand.getDefiningOp();
```

Comment:
```
      auto dequantOp = operand.getDefiningOp<stablehlo::UniformDequantizeOp>();
```

and merge the following two checks.

Reply:
two separate checks is better for logging-granularity which will help in debugging..?

Reply:
Yes the `getDefiningOp<OpTy>` merges the two checks if operand is optype and if defining operation is of `OpTy`. This will simplify the check to "if the operand is coming from a dequant op" as compored to if "operand is defined using an op" and "defined op is deqaunt". Do we need that distinction in the code ?

 Tradeoff cost of keeping the simplified form in the code for someone to understand the failure reason comes down to debugging "why a defined op is not uniform_dequantize", which is not much. 

Again it been a topic of (code simplicity vs debuggability) is subjective, so feel free to make a choice.

Reply:
I unsure either, don't see a major value of keeping separate checks but Pass itself is simple, having two separate checks won't add a lot of complexity.  Decided to keep two separate checks.  Thanks for the discussion. Not a huge tradeoff to consider in this scenario, will be helping in future decisions.